### PR TITLE
madlib-605: We write some C aggregations to improve the performance of C4.5/RF

### PR DESCRIPTION
**Pull request by renyi533**
We write some C aggregations to improve the performance of C4.5/RF. This task is related with these aggregations.

### CODE REVIEW

File: methods/cart/src/pg_gp/dt.c [hunk 1]
```diff
@@ -2126,3 +2339,482 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+    if (rebuild_array)
+    {
+        /* construct a new array to keep the aggr states. */
+        pg_count_array =
+        	construct_array(
+        		(Datum *)count_array,
+                array_length,
+                INT8OID,
+                sizeof(int64),
+                true,
```
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c:2461]: We shouldn't hard-code it, should we?

File: methods/cart/src/pg_gp/dt.c [hunk 2]
```diff
@@ -2126,3 +2339,482 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+ *       The input value will be trimed before it is tested. 
+ *       We define this test simply because a lot of datasets
+ *       use these three representations for missing values before 
+ *       they are loaded into databases. After loading, if these
+ *       strings are not properly translated, the tables will contain 
+ *       '', '?' instead of NULL. 
+ *
+ */
+Datum
+dt_is_miss_value
```
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c]: Hmm, wondering what it is for.  If the special character is needed, why isn't it only NULL?

File: methods/cart/src/pg_gp/dt.c [hunk 3]
```diff
@@ -2126,3 +2339,482 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+			build_state.dvalues[iterator_idx]   = Float8GetDatum(1);
+		}
+
+		/* put the elem into the target slot */
+		build_state.dvalues[elem_idx << 1]       = elem;
+		build_state.dvalues[(elem_idx << 1) + 1] = 
+            Float8GetDatum(PG_ARGISNULL(1) ? 1 : 0);
+		lbs[0] = 1;
+
+		state = construct_md_array
```
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c]: Is this actually construct_array?

Comment (KevinYL) [inline methods/cart/src/pg_gp/dt.c]: Actually, construct_array calls construct_md_array internally.

File: methods/cart/src/pg_gp/dt.c (hunk 1 above)
Comment (renyi533) [inline methods/cart/src/pg_gp/dt.c:2461]: What do you mean here? Actually, the size of the array is not hard-coded. We calculate and store the value to array_length.

File: methods/cart/src/pg_gp/dt.c (hunk 3 above)
Comment (renyi533) [inline methods/cart/src/pg_gp/dt.c]: After investigation, we think "construct_array" cannot handle arrays with null values. Therefore, we directly use construct_md_array here.

File: methods/cart/src/pg_gp/dt.c (hunk 2 above)
Comment (renyi533) [inline methods/cart/src/pg_gp/dt.c]: Many original data sets form UCI or somewhere else use '?' or empty string to denote NULL values. During encoding, we need convert these values to NULL. 

This function is used to identify these fields with null values.

File: methods/cart/src/pg_gp/dt.c (hunk 1 above)
Comment (aojwang) [inline methods/cart/src/pg_gp/dt.c:2461]: I think Hitoshi means that we hard code "pass by value" to true.

Actually, we plan to fix that after the code review done. The reason why we don't fix it here is to separate the performance improvement and bug fixes, so that you can review them easily. If you prefer to fix them in this review, we will do that. Thanks.

Comment (renyi533) [inline methods/cart/src/pg_gp/dt.c:2461]: Yes. Some other components, such as sparse vector, also have this issue. We have a separate dedicated issue number to resolve this.

We think we can submit another pull request to address this issue.

File: methods/cart/src/pg_gp/dt.c (hunk 3 above)
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c]: I don't see you put NULL elements in the array, as you designed it like array[2n] = value[n] and array[2n+1] = isnull[n].

File: methods/cart/src/pg_gp/dt.c [hunk 4]
```diff
@@ -2126,3 +2339,482 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+    dt_check_error_value
+        (
+            elem_cnt > 0,
+            "array_size:%d should be bigger than zero",
+            elem_cnt
+        );
+
+    dt_check_error_value
+        (
+            elem_idx >= 0 && elem_idx < elem_cnt,
```
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c:2551]: You have to check it with the actual array size, not the user input size.

File: methods/cart/src/pg_gp/dt.c [hunk 5]
```diff
@@ -2126,3 +2339,482 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+            "the size of the two state array must be the same"
+        );
+
+	elem_cnt = (ARR_DIMS(arg0)[0]) >> 1;
+
+	for (iterator_idx = 0; iterator_idx < elem_cnt; iterator_idx++)
+	{
+        elem_idx = iterator_idx << 1;
+
+		if (0 == (int)((float8*)ARR_DATA_PTR(arg1))[elem_idx + 1])
```
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c:2681]: Are we sure if arg0[elem_idx] is NULL?  I suggest adding "else" for this condition just in case.

File: methods/cart/src/pg_gp/dt.c [hunk 6]
```diff
@@ -2126,3 +2339,482 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+        );
+
+    dt_check_error_value
+        (
+            ARR_DIMS(state)[0] == (elem_cnt << 1),
+            "The dimension of state array should be %d",
+            (elem_cnt << 1)
+        );
+
+	((float8*)ARR_DATA_PTR(state))[(elem_idx << 1)]     = DatumGetFloat8(elem);
```
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c:2610]: Don't we care if the input has duplicate?  I guess you are overwriting the element specified by the subscript argument, but it is possible that the subscript has been assigned before.

File: methods/cart/src/pg_gp/dt.c [hunk 7]
```diff
@@ -2126,3 +2339,482 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+        result = ('\0' == p[0]) || ('?' == p[0] && 1 ==  strlen(p));
+    }
+
+    PG_RETURN_BOOL(result);
+}
+PG_FUNCTION_INFO_V1(dt_is_miss_value);
+
+
+/*
+ * @brief The step function of the aggregate array_indexed_agg.
```
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c:2504]: Can you explain up to how many elements did you try?  I believe it's not scalable with many of elements in an array, and that's why array_agg doesn't use HashAgg.

File: methods/cart/src/pg_gp/dt.c (hunk 4 above)
Comment (KevinYL) [inline methods/cart/src/pg_gp/dt.c:2551]: At this point we don't know whether the array has been allocated. Checked bellow in this function.

File: methods/cart/src/pg_gp/dt.c (hunk 5 above)
Comment (KevinYL) [inline methods/cart/src/pg_gp/dt.c:2681]: No else need to do. Just taken the non-null one.  And in this environment, the pre-steps make sure there is no duplicates in a given index. 

At the same time we can't detect conditions that a given index has two duplicate nulls.

File: methods/cart/src/pg_gp/dt.c (hunk 3 above)
Comment (KevinYL) [inline methods/cart/src/pg_gp/dt.c]: modified to use consturct_array

File: methods/cart/src/pg_gp/dt.c (hunk 7 above)
Comment (KevinYL) [inline methods/cart/src/pg_gp/dt.c:2504]: It's up to the number of attributes. The most number of elements we tested is about 1500. I want to know why this is not scalable. Using this method, the aggregate process is the same as using ArrayBuildState.

File: methods/cart/src/pg_gp/dt.c (hunk 6 above)
Comment (KevinYL) [inline methods/cart/src/pg_gp/dt.c:2610]: The pre-steps make sure that there is no duplicates.

File: methods/cart/src/pg_gp/dt.c (hunk 4 above)
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c:2551]: I'm suggesting adding the boundary check later than here.

File: methods/cart/src/pg_gp/dt.c (hunk 6 above)
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c:2610]: What is pre-steps?  It would look better if you added the check here to make sure the input is as you expected.

File: methods/cart/src/pg_gp/dt.c (hunk 2 above)
Comment (haradh1) [inline methods/cart/src/pg_gp/dt.c]: What about NULLIF()?

File: methods/cart/src/pg_gp/dt.c (hunk 4 above)
Comment (KevinYL) [inline methods/cart/src/pg_gp/dt.c:2551]: Please see line 2637

File: methods/cart/src/pg_gp/dt.c (hunk 6 above)
Comment (renyi534) [inline methods/cart/src/pg_gp/dt.c:2610]: In previous steps, we can guarantee the input data do not have duplicate index.

File: methods/cart/src/pg_gp/dt.c (hunk 2 above)
Comment (aojwang) [inline methods/cart/src/pg_gp/dt.c]: Done. use the following code.

``` sql
NULLIF(NULLIF(btrim(fval, ' '), '?'), '') as fval
```

File: methods/cart/src/pg_gp/dt.c [hunk 8]
```diff
@@ -2126,3 +2339,438 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+	(
+	PG_FUNCTION_ARGS
+	)
+{
+    ArrayType *pg_count_array    = NULL;
+    int array_dim                = 0;
+    int *p_array_dim             = NULL;
+    int array_length             = 0;
+    int64 *count_array           = NULL;
+    int max_num_of_classes  	 = PG_GETARG_INT32(1);
```
Comment (cwelton) [inline methods/cart/src/pg_gp/dt.c:2373]: And what if the first arg is null?

File: methods/cart/src/pg_gp/dt.c [hunk 9]
```diff
@@ -2126,3 +2339,438 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+    int64  count                 = PG_ARGISNULL(2)?0:PG_GETARG_INT64(2);
+    int    class                 = PG_ARGISNULL(3)?0:PG_GETARG_INT32(3);
+    bool rebuild_array     		 = false;
+
+    dtelog(WARNING, "max_num_of_classes:%d, count:%d, class:%d",
```
Comment (cwelton) [inline methods/cart/src/pg_gp/dt.c]: WARNING?

File: methods/cart/src/pg_gp/dt.c [hunk 10]
```diff
@@ -2126,3 +2339,438 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+			"It must be in range from 0 to the number of classes",
+			class
+		);
+
+    /* test if the first argument (class count array) is null */
+    if (PG_ARGISNULL(0))
+    {
+    	/*
+    	 * We assume the maximum number of classes is limited (up to millions),
+    	 * so that the allocated array won't break our memory limitation.
```
Comment (cwelton) [inline methods/cart/src/pg_gp/dt.c:2399]: You assume that, but don't check it?

File: methods/cart/src/pg_gp/dt.c [hunk 11]
```diff
@@ -2126,3 +2339,438 @@ Datum table_exists(PG_FUNCTION_ARGS)
...
+    		(
+    			array_length == max_num_of_classes,
+    			"dt_acc_count_sfunc invalid array length: %d. "
+    			"The length of class count array must be "
+    			"equal to the total number classes",
+    			array_length
+    		);
+    }
+
+    count_array[class - 1] += count;
```
Comment (cwelton) [inline methods/cart/src/pg_gp/dt.c:2450]: We have guaranteed that class >= 0.  However if class == 0 this will result in memory corruption.

File: methods/cart/src/pg_gp/dt.c (hunk 8 above)
Comment (renyi534) [inline methods/cart/src/pg_gp/dt.c:2373]: I will add more check here.

File: methods/cart/src/pg_gp/dt.c (hunk 9 above)
Comment (renyi534) [inline methods/cart/src/pg_gp/dt.c]: dtelog is a macro which is only present in debug build. In release build, we never enable this macro.

We can also replace WARNING with NOTICE.

File: methods/cart/src/pg_gp/dt.c (hunk 10 above)
Comment (renyi534) [inline methods/cart/src/pg_gp/dt.c:2399]: OK. I will check it.

File: methods/cart/src/pg_gp/dt.c (hunk 11 above)
Comment (renyi533) [inline methods/cart/src/pg_gp/dt.c:2450]: Good catch. We should ensure it is greater than 0 rather than greater than or equal to zero. I will change.

File: methods/cart/src/pg_gp/dt.c (hunk 9 above)
Comment (renyi533) [inline methods/cart/src/pg_gp/dt.c]: I removed the dtelog, which is only useful for debugging purpose.
